File: appendixA/src/main/csharp/Movie/DiscountPolicy.cs
using System;
using Monetary;
using System.Collections.Generic;

namespace MovieReservation
{
    public abstract class DiscountPolicy
    {
        private DiscountCondition[] conditions;

        public PriceDiscountPolicy(params DiscountCondition[] conditions) {
            this.conditions = conditions;
        }

        public Money CalculateDiscountAmount(Screening screening)
        {
			foreach (var each in conditions)
			{
				if (each.IsSatisfiedBy(screening))
				{
					return getDiscountAmount(screening);
				}
			}

			return screening.MovieFee;
		}

		abstract protected Money getDiscountAmount(Screening screening);
	}

    public class AmountDiscountPolicy : DiscountPolicy
	{
    	private Money discountAmount;

        public AmountDiscountPolicy(Money discountAmount, params DiscountCondition[] conditions) : base(conditions)
    	{
    		this.discountAmount = discountAmount;
    	}

    	protected override Money getDiscountAmount(Screening screening)
    	{
    		return discountAmount;
    	}
    }

    public class PercentDiscountPolicy : DiscountPolicy
	{
    	private double percent;

        public PercentDiscountPolicy(double percent, params DiscountCondition[] conditions) : base(conditions)
    	{
    		this.percent = percent;
    	}

    	protected override Money getDiscountAmount(Screening screening)
    	{
    		return screening.MovieFee * percent;
    	}
    }

    public class NoneDiscountPolicy : PriceDiscountPolicy
	{
    	protected override Money getDiscountAmount(Screening screening)
    	{
    		return new Money(0);
    	}
    }
}

File: chapter13/src/main/java/org/eternity/movie/step01/DiscountPolicy.java
package org.eternity.movie.step01;

import org.eternity.money.Money;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public abstract class DiscountPolicy {
    private List<DiscountCondition> conditions = new ArrayList<>();

    public DiscountPolicy(DiscountCondition ... conditions) {
        this.conditions = Arrays.asList(conditions);
    }

    public Money calculateDiscountAmount(Screening screening) {
        checkPrecondition(screening);

        Money amount = Money.ZERO;
        for(DiscountCondition each : conditions) {
            if (each.isSatisfiedBy(screening)) {
                amount = getDiscountAmount(screening);
                checkPostcondition(amount);
                return amount;
            }
        }

        amount = screening.getMovieFee();
        checkPostcondition(amount);
        return amount;
    }

    protected void checkPrecondition(Screening screening) {
        assert screening != null &&
                screening.getStartTime().isAfter(LocalDateTime.now());
    }

    protected void checkPostcondition(Money amount) {
        assert amount != null && amount.isGreaterThanOrEqual(Money.ZERO);
    }


    abstract protected Money getDiscountAmount(Screening Screening);
}
